File: abc396/e/main.py
from collections import defaultdict, deque
from itertools import combinations, permutations
from bisect import bisect, bisect_left, bisect_right
from sortedcontainers import SortedSet, SortedDict, SortedList
import math
import sys
sys.set_int_max_str_digits(10000000)
sys.setrecursionlimit(4100000)
import heapq
try:
    from icecream import ic
except ImportError:  # Graceful fallback if IceCream isn't installed.
    ic = lambda *a: None if not a else (a[0] if len(a) == 1 else a)
# def error(*args, end="\n"): print("[stderr]", *args, end=end, file=sys.stderr)
MOD = 998244353
INF = float("inf")
MINF = -float("inf")

N, M = map(int, input().split())
# A = list(map(int, input().split()))

L = []
for _ in range(M):
    x, y, z = map(int, input().split())
    x, y = x-1, y-1
    L.append((x, y, z))

ans = [[-1]*N for _ in range(30)]

for i in range(30):
    G = [[] for _ in range(N)]
    for x, y, z in L:
        if (z >> i) & 1:
            G[x].append((y, 1))
            G[y].append((x, 1))
        else:
            G[x].append((y, 0))
            G[y].append((x, 0))
    group_lists = []
    seen = set()
    for j in range(N):
        if j in seen:
            continue
        groups = [set(), set()]
        seen.add(j)
        Q  = deque()
        groups[0].add(j)
        Q.append((j, 0))
        while Q:
            v, _group = Q.popleft()
            for u, w in G[v]:
                if w == 0: # 同じグループ
                    if u in seen:
                        if u in groups[1-_group]:
                            print(-1)
                            sys.exit()
                        continue
                    groups[_group].add(u)
                    seen.add(u)
                    Q.append((u, _group))
                else:
                    if u in seen:
                        if u in groups[_group]:
                            print(-1)
                            sys.exit()
                        continue
                    groups[1-_group].add(u)
                    seen.add(u)
                    Q.append((u, 1-_group))
        if len(groups[0]) < len(groups[1]):
            groups[0], groups[1] = groups[1], groups[0]
        group_lists.append(groups)

    for group0, group1 in group_lists:
        for v in group0:
            ans[i][v] = 0
        for v in group1:
            ans[i][v] = 1

ret = [0]*N
for i in range(30):
    for j in range(N):
        ret[j] |= (ans[i][j] << i)

print(*ret)

File: abc396/e/main.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 998244353;
const long long INF = 1LL << 60;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int N, M;
    cin >> N >> M;

    vector<tuple<int, int, int>> L;
    for (int i = 0; i < M; ++i) {
        int x, y, z;
        cin >> x >> y >> z;
        --x; --y; // 0-indexed
        L.emplace_back(x, y, z);
    }

    vector<vector<int>> ans(30, vector<int>(N, -1));

    for (int i = 0; i < 30; ++i) {
        vector<vector<pair<int, int>>> G(N);
        for (auto &[x, y, z] : L) {
            int bit = (z >> i) & 1;
            G[x].emplace_back(y, bit);
            G[y].emplace_back(x, bit);
        }

        vector<bool> seen(N, false);
        vector<pair<set<int>, set<int>>> group_lists;

        for (int j = 0; j < N; ++j) {
            if (seen[j]) continue;

            set<int> groups[2];
            queue<pair<int, int>> Q;
            Q.emplace(j, 0);
            groups[0].insert(j);
            seen[j] = true;

            while (!Q.empty()) {
                auto [v, g] = Q.front(); Q.pop();
                for (auto &[u, w] : G[v]) {
                    if (w == 0) {
                        if (groups[1 - g].count(u)) {
                            cout << -1 << '\n';
                            return 0;
                        }
                        if (!seen[u]) {
                            groups[g].insert(u);
                            seen[u] = true;
                            Q.emplace(u, g);
                        }
                    } else {
                        if (groups[g].count(u)) {
                            cout << -1 << '\n';
                            return 0;
                        }
                        if (!seen[u]) {
                            groups[1 - g].insert(u);
                            seen[u] = true;
                            Q.emplace(u, 1 - g);
                        }
                    }
                }
            }

            // swap if needed
            if (groups[0].size() < groups[1].size())
                swap(groups[0], groups[1]);

            group_lists.emplace_back(groups[0], groups[1]);
        }

        for (auto &[group0, group1] : group_lists) {
            for (int v : group0) ans[i][v] = 0;
            for (int v : group1) ans[i][v] = 1;
        }
    }

    vector<int> ret(N, 0);
    for (int i = 0; i < 30; ++i) {
        for (int j = 0; j < N; ++j) {
            if (ans[i][j] == 1)
                ret[j] |= (1 << i);
        }
    }

    for (int i = 0; i < N; ++i) {
        cout << ret[i] << (i == N - 1 ? '\n' : ' ');
    }

    return 0;
}
